### PR TITLE
[yang]: modify Vlan Member to include PortChannel along with Port

### PR DESCRIPTION
**Pull request by ArthiSivanantham**
Signed-off-by: Arthi Sivanantham <arthi_sivanantham@dell.com>

#### Why I did it
To include PortChannel as Vlan Member (in addition to the already existing physical port)

#### How I did it
Modified the type of vlan member port to a union of PortChannel name and Port name.

#### How to verify it
sonic_yang_models package build
sonic_yang_mgmt package build

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

#### Description for the changelog
Modify Vlan Member to include PortChannel along with Port

### CODE REVIEW

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: please do not change the name here.

**Comment by lguohan** [comment]
@praveen-li to review

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: In community call regarding naming, we have decided to name it as port for physical ethernet port and ifname for mix of ports cases. And just “name” for interfaces defined in sonic-port.yang, sonic-portchannel.yang, sonic-vlan.yang. Also since this is a key attribute, changing attribute-name will not cause any backward compatibility impact. Based on these, we have updated here

Comment (praveen-li) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: ifname for mix of ports cases.?, ..., ifname is for interface name not for mix of ports.

We can stick to port:
as https://github.com/Azure/sonic-buildimage/blob/29763cd095e8bb63c0756a84b375ef5d445f4648/src/sonic-yang-models/yang-models/sonic-acl.yang#L258

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:19]: If you want to fix the tab and space issue, I will suggest raising a PR that changes just tab to space. Send me the link I will quickly approve that.  
And then on top of that raise this PR.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/files/sample_config_db.json:627]: Changes in this file looks good.

File: src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json:22]: Must be LeafRef only?,

File: src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json:29]: Must be LeafRef

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (lguohan) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: i do not understand this. vlan_member is only for one member port, why do we have a list of port now?

Comment (rathnasabapathyv) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: @lguohan : Same VLAN can have multiple ports as members. Also same VLAN can have ethernet port & port-channel as members here. The comment mix-of-ports is mentioned in the perspective of different types of interfaces (Ethernet, Port-channel). Hope this clarifies the context of discussion.

@praveen-li : I remember, while we were discussing about consistent naming for interface names for this PR (https://github.com/Azure/sonic-buildimage/pull/7049) in call, we said port is generally in the context specific to  "Physical ethernet interfaces" and ifname (since it's interface-name) can be widely used to refer multiple different types of interfaces like Ethernet & Port-channel together + other types of interfaces like VLAN, loopback in other applications context like BGP where it will enable on all types of interfaces. In those understanding, here we have renamed it as ifname to include Etherent & Port-channel together.

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: I've reverted to the usage of tabs. Let's complete this PR now, only with the required changes (for PortChannel as Vlan member)

File: src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json
Comment (ArthiSivanantham) [reply src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json:22]: The type of vlan member (ifname) is changed from "leafref" to a union of 2 leafrefs (port and portchannel). Hence an "InvalidValue" error is returned from libyang. 

In the existing TC (with LeafRef), seeing this build error:
------------------- Test 59: Configure a member port in VLAN_MEMBER table which does not exist.---------------------
libyang[0]: Invalid value "Ethernet156" in "ifname" element. (path: /sonic-vlan:sonic-vlan/VLAN_MEMBER/VLAN_MEMBER_LIST[name='Vlan100'][ifname='Ethernet156']/ifname)
INFO:YANG-TEST:Invalid value "Ethernet156" in "ifname" element.
**ERROR:YANG-TEST: Exception >Mismatch ['Leafref', 'non-existing'] and Invalid value "Ethernet156" in "ifname" element.< in /sonic/src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py:211**
INFO:YANG-TEST:Configure a member port in VLAN_MEMBER table which does not exist. Failed

Since the type is changed, I've modified the "eStrKey" appropriately.

File: src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json
Comment (ArthiSivanantham) [reply src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json:29]: Same explanation as above.

**Comment by rathnasabapathyv** [comment]
@praveen-li : @ArthiSivanantham has addressed few comments with changes & we have responded all other comments. Can you please check & approve the PR, if no further comments? Thanks.

File: src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json
Comment (praveen-li) [reply src/sonic-yang-models/tests/yang_model_tests/tests/vlan.json:22]: okie, sure if Libyang behaves this way...they we can use Invalid value only

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:3]: We can remove this as of now.

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:199]: We can remove these changes also as of now.

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: Addressed

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: Addressed

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: Addressed

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [reply src/sonic-yang-models/yang-models/sonic-vlan.yang]: Addressed

**Comment by ArthiSivanantham** [comment]
@praveen-li - All comments are addressed. Can you please review and approve the PR ?

**Comment by ArthiSivanantham** [comment]
@lguohan - Can you please check this generic build failure ? 

build_error

File: src/sonic-yang-models/yang-models/sonic-vlan.yang
Comment (ArthiSivanantham) [inline src/sonic-yang-models/yang-models/sonic-vlan.yang:180]: This leaf includes a custom validation (sonic-ext:custom-validation ValidateVlanMember;), which will be added once the support for custom validation is merged via' this PR,
https://github.com/Azure/sonic-buildimage/pull/6917

**Comment by rathnasabapathyv** [comment]
@ArthiSivanantham : I see a recent conversation in sonic-community email thread about this recent compilation errors with sonic-acl.yang & sonic-device_metadata.yang. Looks like it's fixed in latest master. Can you try to sync the latest master to your branch and see whether builds succeeds?

**That community email conversation about this:**

image

**Comment by praveen-li** [comment]
We need this PR as well to fix it, thx.

https://github.com/Azure/sonic-buildimage/pull/7436




Thanks,

Regards
Praveen
________________________________

**Comment by ArthiSivanantham** [comment]
@praveen-li - All review comments are addressed. Can you please review and approve this PR ?

**Review by rathnasabapathyv — APPROVED** [review]

**Review by lguohan — APPROVED** [review]

**Pull request merged**
